include/boost/multiprecision/integer.hpp

```diff
@@ -190,50 +190,75 @@ BOOST_MP_CXX14_CONSTEXPR typename std::enable_if<boost::multiprecision::detail::
 }
 
 template <class Integer>
-BOOST_MP_CXX14_CONSTEXPR typename std::enable_if<boost::multiprecision::detail::is_integral<Integer>::value, Integer>::type sqrt(const Integer& x, Integer& r)
+BOOST_MP_CXX14_CONSTEXPR typename std::enable_if<boost::multiprecision::detail::is_integral<Integer>::value, Integer>::type karatsuba_sqrt(const Integer& x, Integer& r, size_t bits)
 {
-   //
-   // This is slow bit-by-bit integer square root, see for example
-   // http://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Binary_numeral_system_.28base_2.29
-   // There are better methods such as http://hal.inria.fr/docs/00/07/28/54/PDF/RR-3805.pdf
-   // and http://hal.inria.fr/docs/00/07/21/13/PDF/RR-4475.pdf which should be implemented
-   // at some point.
-   //
-   Integer s = 0;
-   if (x == 0)
-   {
-      r = 0;
-      return s;
+#ifndef BOOST_MP_NO_CONSTEXPR_DETECTION
+   // std::sqrt is not constexpr by standard, so use this 
+   if (BOOST_MP_IS_CONST_EVALUATED(bits)) {
+      if (bits <= 4) {
+         if (x == 0) {
+            r = 0;
+            return 0;
+         }
+         else if (x < 4) {
+            r = x - 1;
+            return 1;
+         }
+         else if (x < 9) {
+            r = x - 4;
+            return 2;
+         }
+         else {
+            r = x - 9;
+            return 3;
+         }
+      }
    }
-   int g = msb(x);
-   if (g == 0)
-   {
-      r = 1;
-      return s;
+   else
+#endif
+   // we can calculate it faster with std::sqrt
+   if (bits <= 64) {
```

Comment:
If you felt so inclined to do so... it might pay to check which is fastest here:

* double precision sqrt and 53 bits.
* long double and 64 bits, but now using the old x87 co-processor.
* __float128 and 113 bits (when supported).

There is also an assumption here that long double has a 64-bit mantissa, but 53 and 113 bit mantissa's are also common as I'm sure you know.

Reply:
Might it make sense to check? :

```
if(bits <= std::numeric_limits<double>::digits)
```

Reply:
... and then use `sqrt` from `<cmath>` if that is your inclination for base-case square root?